lib/logreader.c

```diff
@@ -501,11 +506,13 @@ log_reader_fetch_log(LogReader *self)
 
   if ((self->options->flags & LR_IGNORE_AUX_DATA))
     aux = NULL;
-
   log_transport_aux_data_init(aux);
+
   if (self->handshake_in_progress)
     {
-      return log_reader_process_handshake(self);
+      gboolean succ = log_reader_process_handshake(self);
+      if (FALSE == succ || self->handshake_in_progress)
+        return FALSE;
```

Comment:
This what is really needed here.
This check must not return true ever.
If the handshake is needed and it fails or not completed yet, it should return false.
If the handshake is not needed or it is finised fine already, the flow should go forward and finish the flush operation, otherwise this flush will be lost (as the caller might think its flush request is processed and will not attempt it once again).
The latter can be experienced currently, just try my example bellow.